executor/src/witgen/mod.rs

```diff
@@ -111,7 +111,7 @@ impl<T: FieldElement> WitgenCallbackContext<T> {
         current_witness: &[(String, Vec<T>)],
         challenges: BTreeMap<u64, T>,
         stage: u8,
-    ) -> Vec<(String, Vec<T>)> {
+    ) -> (Witness<T>, Publics<T>) {
```

Comment:
Maybe not a topic of this PR, but I think this would need to receive the stage-0 publics as input as well, no? For example, we could have a constraint `stage1_public = stage0_public;`, or write a stage-0 public to the bus, or something like that.

Perhaps it makes sense to do that as part of this PR too if you implement [this suggestion](https://github.com/powdr-labs/powdr/pull/2556#discussion_r2007332681). Basically, we'd always group the trace and the publics-